examples/nlp/language_modeling/megatron_change_num_partitions_pp.py

```diff
@@ -0,0 +1,753 @@
+            )
+            continue
+
+        # Log some useful comparison of tensors that are being mapped.
+        # Note that the global param index for layers and modules may be different but the shapes
+        # and semantics of the layer should match.
+        if DEBUG_PRINT:
+            print(
+                "Index:", idx, "Model param  :", param_name, param.shape,
+            )
+            print(
+                "Index:", idx, "Global params:", partitions[1][idx], partitions[0][idx].shape,
+            )
+
+        # Tensor Parallel Splitting
+        split = compute_tp_splits(
+            param_name, param, partitions, idx, tp_size, pp_size, pp_rank, pp_split_rank, megatron_legacy
+        )
+
+        splits.append(split)
+        idx += 1
+
+        # When pp split rank is equal to current pp rank, we need to first inject the encoder embeddings
+        # and then reset the index to the originally computed index
+        if enc_dec_share_token_embeddings_count > 0:
+            if enc_dec_share_token_embeddings_count - 1 == 0:
+                idx = computed_index
+
+            enc_dec_share_token_embeddings_count -= 1
+
+    # Inject the EncDec shared embeddings intermediate tensor
+    # at one random location in the decoder of this TP PP rank.
+    # Note that usually it is the second last tensor, but to avoid specific index we search for it
+    # again.
+    if shared_enc_dec_embeddings_intermediate:
+        for param_name, param in model.named_parameters():
+            if param_name == 'enc_dec_model.word_embeddings.weight':
+                print("Found intermediate shared embedding, injecting")
+                split = compute_tp_splits(
+                    param_name, param, partitions, 0, tp_size, pp_size, pp_rank, pp_split_rank, megatron_legacy
+                )
+                splits.insert(intermediate_shared_embedding_location, split)
+                break
+
+    # Compute the new offset for the next PP rank in reverse order
+    # Add 1 to offset to account for last PP rank's duplicated Embedding
+    offset_diff = offset - num_params
+    # GPT offset correction
+    if pp_size > 1 and pp_rank == (pp_size - 1) and pp_split_rank == 0:
+        offset_diff += 1
+    # T5 offset correction for shared embedding when pp split rank == pp rank
+    if shared_enc_dec_embeddings:
+        offset_diff += 2
+    # T5 offset correction for intermediate shared embedding when pp rank > pp split rank
+    if shared_enc_dec_embeddings_intermediate:
+        offset_diff += 1
+
+    # Finalize the new offset
+    new_offset = offset_diff
+
+    # Save each of the TP ranks in reverse order
+    # This is done so that the last PP rank will save the last TP rank only after all other PP TP ranks are saved
+    # The final rank will then save a new NeMo file with all other ranks inside.
+    for tp_rank in range(tp_size - 1, -1, -1):
```

Comment:
nice work around for saving the nemo files with one worker.

Reply:
@aklife97 came up with this for TP change, turns out its needed for PP change too :D